[ingestion/data-quality issue] SLSA parser should ensure fields are non-null before dereferencing

**Describe the bug**

Currently [SLSA parser](https://github.com/guacsec/guac/blob/main/pkg/ingestor/parser/slsa/parser_slsa.go#L199) sometimes assumes fields are non-nil and dererefences them, which when provided with a non-complete document causes a NPE.



**To Reproduce**

Make a SLSA predicate where there is no "predicate" field

**Expected behavior**
It should fail via throwing an error when information required is not present. or alter the inputs to GQL accordingly

**Screenshots**
```
{"level":"info","ts":1689609405.540687,"caller":"parser/parser.go:130","msg":"parsing document tree with root type: SLSA"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1fa086f]

goroutine 149 [running]:
github.com/guacsec/guac/pkg/ingestor/parser/slsa.fillSLSA01(...)
	/Users/lumb/go/src/github.com/guacsec/guac/pkg/ingestor/parser/slsa/parser_slsa.go:199
github.com/guacsec/guac/pkg/ingestor/parser/slsa.(*slsaParser).getSLSA(0xc001fa4380)
	/Users/lumb/go/src/github.com/guacsec/guac/pkg/ingestor/parser/slsa/parser_slsa.go:235 +0xef
github.com/guacsec/guac/pkg/ingestor/parser/slsa.(*slsaParser).Parse(0xc001fa4380, {0xc0003e1fb0?, 0x250f72c?}, 0x4?)
	/Users/lumb/go/src/github.com/guacsec/guac/pkg/ingestor/parser/slsa/parser_slsa.go:82 +0x73
github.com/guacsec/guac/pkg/ingestor/parser.parseHelper({0x2792fa0, 0xc0002e7940}, 0xc0006a4240)
	/Users/lumb/go/src/github.com/guacsec/guac/pkg/ingestor/parser/parser.go:190 +0x83
github.com/guacsec/guac/pkg/ingestor/parser.(*docTreeBuilder).parse(0xc000629a48, {0x2792fa0, 0xc0002e7940}, 0xc0013fa260, 0x0?)
	/Users/lumb/go/src/github.com/guacsec/guac/pkg/ingestor/parser/parser.go:161 +0x5c
github.com/guacsec/guac/pkg/ingestor/parser.ParseDocumentTree({0x2792fa0, 0xc0002e7940}, 0xc0013fa260)
	/Users/lumb/go/src/github.com/guacsec/guac/pkg/ingestor/parser/parser.go:131 +0x19c
github.com/guacsec/guac/cmd/guacone/cmd.getIngestor.func1(0x2785a80?)
	/Users/lumb/go/src/github.com/guacsec/guac/cmd/guacone/cmd/files.go:255 +0x27
github.com/guacsec/guac/cmd/guacone/cmd.glob..func5.1.1()
	/Users/lumb/go/src/github.com/guacsec/guac/cmd/guacone/cmd/files.go:169 +0x2ef
golang.org/x/sync/errgroup.(*Group).Go.func1()
	/Users/lumb/go/pkg/mod/golang.org/x/sync@v0.3.0/errgroup/errgroup.go:75 +0x64
created by golang.org/x/sync/errgroup.(*Group).Go
	/Users/lumb/go/pkg/mod/golang.org/x/sync@v0.3.0/errgroup/errgroup.go:72 +0xa5
```


**GUAC version**
Are you using a released version of GUAC or one build from a commit in the
repository?

**Ingested document(s)**

Can you share the documents that are used to reproduce the ingestion errors or showcase the data quality issues.

**Additional context**
Add any other context about the problem here.

--- a/pkg/ingestor/parser/slsa/parser_slsa.go
+++ b/pkg/ingestor/parser/slsa/parser_slsa.go
@@ -18,6 +18,7 @@ package slsa
 import (
 	"context"
 	"encoding/json"
+	"errors"
 	"fmt"
 	"strings"
 
@@ -39,6 +40,9 @@ import (
 // - A artifact for each digest information
 // - a pkg or source depending on what is represented by the name/URI
 // - An IsOccurence input spec which will generate a predicate for each occurence
+
+var ErrMetadataNil = errors.New("SLSA01 Metadata is nil")
+
 type slsaEntity struct {
 	artifacts []*model.ArtifactInputSpec
 	occurence *model.IsOccurrenceInputSpec
@@ -194,14 +198,20 @@ func getSlsaEntity(name string, digests scommon.DigestSet) (*slsaEntity, error) 
 	return nil, fmt.Errorf("%w unable to get Guac Generic Purl, this should not happen", err)
 }
 
-func fillSLSA01(inp *model.SLSAInputSpec, stmt *in_toto.ProvenanceStatementSLSA01) {
+func fillSLSA01(inp *model.SLSAInputSpec, stmt *in_toto.ProvenanceStatementSLSA01) error {
 	inp.BuildType = stmt.Predicate.Recipe.Type
+
+	if stmt.Predicate.Metadata == nil {
+		return ErrMetadataNil
+	}
 	if stmt.Predicate.Metadata.BuildStartedOn != nil {
 		inp.StartedOn = stmt.Predicate.Metadata.BuildStartedOn
 	}
 	if stmt.Predicate.Metadata.BuildFinishedOn != nil {
-		inp.FinishedOn = stmt.Predicate.Metadata.BuildStartedOn
+		inp.FinishedOn = stmt.Predicate.Metadata.BuildFinishedOn
 	}
+
+	return nil
 }
 
 func fillSLSA02(inp *model.SLSAInputSpec, stmt *in_toto.ProvenanceStatementSLSA02) {
@@ -232,7 +242,9 @@ func (s *slsaParser) getSLSA() error {
 	var pred any
 	switch s.header.PredicateType {
 	case slsa01.PredicateSLSAProvenance:
-		fillSLSA01(inp, s.s01smt)
+		if err := fillSLSA01(inp, s.s01smt); err != nil {
+			return fmt.Errorf("could not fill SLSA01: %w", err)
+		}
 		pred = s.s01smt.Predicate
 	case slsa02.PredicateSLSAProvenance:
 		fillSLSA02(inp, s.s02smt)

--- a/pkg/ingestor/parser/slsa/parser_slsa_test.go
+++ b/pkg/ingestor/parser/slsa/parser_slsa_test.go
@@ -18,10 +18,16 @@ package slsa
 import (
 	"context"
 	"testing"
+	"time"
+
+	slsa01 "github.com/in-toto/in-toto-golang/in_toto/slsa_provenance/v0.1"
+
+	"github.com/in-toto/in-toto-golang/in_toto"
 
 	"github.com/google/go-cmp/cmp"
 	"github.com/guacsec/guac/internal/testing/testdata"
 	"github.com/guacsec/guac/pkg/assembler"
+	model "github.com/guacsec/guac/pkg/assembler/clients/generated"
 	"github.com/guacsec/guac/pkg/handler/processor"
 	"github.com/guacsec/guac/pkg/logging"
 )
@@ -63,6 +69,70 @@ func Test_slsaParser(t *testing.T) {
 			//fmt.Println(preds.HasSlsa[0].HasSlsa.SlsaPredicate)
 			if d := cmp.Diff(tt.wantPredicates, preds, testdata.IngestPredicatesCmpOpts...); len(d) != 0 {
 				t.Errorf("slsa.GetPredicate mismatch values (+got, -expected): %s", d)
+			}
+		})
+	}
+}
+
+func Test_fillSLSA01(t *testing.T) {
+	startTime := time.Now()
+	endTime := time.Now()
+	type args struct {
+		inp  *model.SLSAInputSpec
+		stmt *in_toto.ProvenanceStatementSLSA01
+	}
+	tests := []struct {
+		name string
+		args args
+		err  error
+	}{
+		{
+			name: "default",
+			args: args{
+				inp: &model.SLSAInputSpec{},
+				stmt: &in_toto.ProvenanceStatementSLSA01{
+					Predicate: slsa01.ProvenancePredicate{
+						Metadata: &slsa01.ProvenanceMetadata{
+							BuildStartedOn:  &startTime,
+							BuildFinishedOn: &endTime,
+						},
+						Recipe: slsa01.ProvenanceRecipe{
+							Type: "test",
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "stmt predicate metadata is nil",
+			args: args{
+				inp:  &model.SLSAInputSpec{},
+				stmt: &in_toto.ProvenanceStatementSLSA01{},
+			},
+			err: ErrMetadataNil,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			err := fillSLSA01(test.args.inp, test.args.stmt)
+			if err != test.err {
+				t.Fatalf("fillSLSA01() error = %v, expected error %v", err, test.err)
+			}
+			if err != nil {
+				// if the error is equal to the expected error, we can return
+				// we know that the error is equal to the expected error because we checked if it wasn't equal above
+				// if err == nil then we know that the function didn't throw an error, and we have to check the values
+				return
+			}
+
+			if test.args.inp.BuildType != test.args.stmt.Predicate.Recipe.Type {
+				t.Errorf("fillSLSA01() inp.BuildType not equal to stmt.Predicate.Recipe.Type")
+			}
+			if test.args.inp.StartedOn != test.args.stmt.Predicate.Metadata.BuildStartedOn {
+				t.Errorf("fillSLSA01() inp.BuildStartedOn not equal to stmt.Predicate.Metadata.BuildStartedOn")
+			}
+			if test.args.inp.FinishedOn != test.args.stmt.Predicate.Metadata.BuildFinishedOn {
+				t.Errorf("fillSLSA01() inp.BuildFinishedOn not equal to stmt.Predicate.Metadata.BuildFinishedOn")
 			}
 		})
 	}

